flow/layers/layer_state_stack.cc

```diff
@@ -0,0 +1,337 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/layers/layer_state_stack.h"
+#include "flutter/flow/paint_utils.h"
+
+namespace flutter {
+
+void LayerStateStack::setCanvasDelegate(SkCanvas* canvas) {
+  if (canvas_) {
+    canvas_->restoreToCount(canvas_restore_count_);
+    canvas_ = nullptr;
+  }
+  if (canvas) {
+    canvas_restore_count_ = canvas->getSaveCount();
+    canvas_ = canvas;
+    for (auto& state : state_stack_) {
+      if (!state->is_backdrop_filter()) {
+        // BackdropFilter is only applied on the canvas that
+        // was present at the time that the backdrop layer was
+        // first encountered to avoid applying it on every
+        // platform view embedder overlay.
+        state->apply(canvas, nullptr);
+      }
+    }
+  }
+}
+
+void LayerStateStack::setBuilderDelegate(DisplayListBuilder* builder) {
+  if (builder_) {
+    builder_->restoreToCount(builder_restore_count_);
+    builder_ = nullptr;
+  }
+  if (builder) {
+    builder_restore_count_ = builder->getSaveCount();
+    builder_ = builder;
+    for (auto& state : state_stack_) {
+      if (!state->is_backdrop_filter()) {
```

Comment:
Where we apply the BF?

Reply:
Only when it is first pushed onto the state stack. If we update the canvas to a new overlay, we don't want to re-blur everything that is behind that overlay...(do we?)...

Reply:
Oops, but we do need to do a saveLayer so that the restores match up...

Reply:
I swiched from an "is_bdf" conditional to calling a reapply method here in the delegate swap phase. That way, only BDFEntry has a different implementation that does a saveLayer without its backdrop filter.